spring-cloud-deployer-spi/src/main/java/org/springframework/cloud/deployer/spi/app/observation/DefaultAppDeployerObservationConvention.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2013-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.deployer.spi.app.observation;
+
+import java.util.Map;
+
+import io.micrometer.common.KeyValues;
+
+import org.springframework.cloud.deployer.spi.core.AppDeploymentRequest;
+import org.springframework.cloud.deployer.spi.core.RuntimeEnvironmentInfo;
+import org.springframework.lang.Nullable;
+import org.springframework.util.StringUtils;
+
+public class DefaultAppDeployerObservationConvention implements AppDeployerObservationConvention {
+
+	// TODO: Check what is high and what is low cardinality
```

Comment:
Good idea. What threshold is considered low vs. high cardinality these days?

Reply:
Jonatan has created a `HighCardinalityTagsDetector` class that uses the following logic to calculate that an entry goes from low to high cardinality

```java
private static long calculateThreshold() {
        // 10% of the heap in MiB
        long allowance = Runtime.getRuntime().maxMemory() / 1024 / 1024 / 10;

        // 2k Meters can take ~1MiB, 2M Meters can take ~1GiB
        return Math.max(1_000, Math.min(allowance * 2_000, 2_000_000));
    }
```